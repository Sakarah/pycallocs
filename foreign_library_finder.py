import allocs
import importlib
import importlib.abc
import importlib.machinery
import sys

# These can be modified by the user to change the finder search paths
lib_extension = ".so"

# The blank search path has a special meaning as dlopen is using its own search 
# paths when the file name does not contain any /
search_paths = [""]

class ForeignLibraryFinder(importlib.abc.MetaPathFinder):
    """
        Meta path finder searching foreign libraries usable
        with allocs.ForeignLibraryLoader.
    """

    def find_spec(fullname, path, target):
        if path is not None:
            # Should we support submodules ?
            return None
        
        for base_path in search_paths:
            try:
                filename = base_path + fullname + lib_extension
                loader = allocs.ForeignLibraryLoader(filename)
                return importlib.machinery.ModuleSpec(fullname, loader, origin=filename)
            except ImportError:
                continue

        return None

sys.meta_path.append(ForeignLibraryFinder)
