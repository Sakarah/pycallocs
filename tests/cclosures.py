import foreign_library_finder
foreign_library_finder.search_paths.append("libs/")
import closures

def hi():
    print("Hi")

def add_one(i):
    return i+1

closures.repeat(3, closures.__FUN_FROM___FUN_TO_void(hi))

add_one_closure = closures.__FUN_FROM___ARG0_int_32__FUN_TO_int_32(add_one)
print(closures.fold_int(5, add_one_closure))
print(closures.fold_int(42, add_one_closure))

print(closures.fold_int(10, closures.__FUN_FROM___ARG0_int_32__FUN_TO_int_32(lambda x: (x*2)+1)))
