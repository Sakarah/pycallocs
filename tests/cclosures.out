Hi
Hi
Hi
5
42
1023
